### ISSUE TITLE
Internal error caused by comment inside dict comprehension

### ISSUE THREAD

**Issue by timmwagener** [psf/black#1211]
**Describe the bug** A comment inside a dict comprehension causes black to raise an internal error. The problem seems reproducable via [black playground](https://black.now.sh/?version=master&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4AFWANJdAD2IimZxl1N_WlbvK5V8iDHdFLCSAG5pDGDwk_6MDa4u0LdJ-k5L7C5cYVOPQPBynCkXQ2mnYJfRCUXEzutee0lFMyoZllU4_cvw9xnSkoDuA7hIt-V52mO0memBS51Dgbdr7BTO1xwUcrUwoRX3aXjzukkVWsaOE9q6qCsd0NPY4Ib7_nJ4Xlex0UmscXHVB4WJ_gUvHGzm3VHG_QvtIo86OgWoJ9NPu2WC5Dor6GNK3NQ4B63MtnGZfkY8LWHjyro7Rqc3B6GC84HQVPMyr1toIAAAAKda25l2zbTHAAHuAdcCAADR0nxqscRn-wIAAAAABFla).

**Code** 
```python
# converter.py
def convert(collection):

    # replace all variables by integers
    replacement_dict = {
        variable: f"{index}"
        for index, variable
        # 0 is reserved as line terminator
        in enumerate(collection.variables(), start=1)
    }
```
**Error**
```python
def convert(collection):

    # replace all variables by integers
    replacement_dict = {
        variable: f"{index}"
        for index, variable# 0 is reserved as line terminator in enumerate(
            collection.variables(), start=1
        )
    }

```

**Log** 
```
...
error: cannot format D:\path\to\converter.py
: INTERNAL ERROR: Black produced invalid code: invalid syntax (<unknown>, line 55). Please report a bug on https://github.com/psf/blac
k/issues.  This invalid output might be helpful: C:\Users\XXX\AppData\Local\Temp\blk_1_om2a5y.log
...
```

**To Reproduce** Run black with the total defaults _(`black directory/`)_ on a file that contains the code above.

**Expected behavior** Produce valid code when formatting dict comprehensions with comments inside.

**Environment (please complete the following information):**

- Version: `black==19.10b0`
- OS and Python version: `Win7x64/Python 3.7.5 (tags/v3.7.5:5c02a39a0b, Oct 15 2019, 00:11:34) [MSC v.1916 64 bit (AMD64)] on win32`

**Does this bug also happen on master?** [Yes](https://black.now.sh/?version=master&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4AFWANJdAD2IimZxl1N_WlbvK5V8iDHdFLCSAG5pDGDwk_6MDa4u0LdJ-k5L7C5cYVOPQPBynCkXQ2mnYJfRCUXEzutee0lFMyoZllU4_cvw9xnSkoDuA7hIt-V52mO0memBS51Dgbdr7BTO1xwUcrUwoRX3aXjzukkVWsaOE9q6qCsd0NPY4Ib7_nJ4Xlex0UmscXHVB4WJ_gUvHGzm3VHG_QvtIo86OgWoJ9NPu2WC5Dor6GNK3NQ4B63MtnGZfkY8LWHjyro7Rqc3B6GC84HQVPMyr1toIAAAAKda25l2zbTHAAHuAdcCAADR0nxqscRn-wIAAAAABFla)

**Comment by hauntsaninja** [comment]
I spent some time looking into this. It seems like there's a failure of `append_safe` to actually be safe.

Here's a slightly smaller repro:
```
{
    i: i
    for i
    # a comment
    in range(5)
}
```
This appears to affect lambdas too:
```
(
    lambda z
    # a comment
    : 5
)
```

Not quite sure what a good way to fix this is — I'm fairly new to black code — but if I figure out something I like I'll open a PR.

**Comment by timmwagener** [comment]
:+1:

**Comment by cooperlees** [comment]
Is this still reproducible on top of trunk?

**Comment by ichard26** [comment]
Yes, this bug is still present in master:
```console
(black) richard-26@ubuntu-laptop:~/programming/black$ black test.py --diff --color
error: cannot format test.py: INTERNAL ERROR: Black produced invalid code: invalid syntax (<unknown>, line 6). Please report a bug on https://github.com/psf/black/issues.  This invalid output might be helpful: /tmp/blk_1gc5ml17.log
Oh no! 💥 💔 💥
1 file would fail to reformat.
```

For further reference, this issue was closed as it was considered as a duplicate of #563 which is also reproducible ATM.
